### ISSUE TITLE
Implement `cv::threshold` with its variants

### ISSUE THREAD

**Issue by edgarriba** [kornia/kornia#1804]
## 🚀 Feature

We would like to have in `kornia.enhance` a batched device agnostic version of the `cv::threshold` including the different variants.

Different threshold types:
- [ ] THRESH_BINARY
- [ ] THRESH_BINARY_INV
- [ ] THRESH_TRUNC
- [ ] THRESH_TOZERO
- [ ] THRESH_TOZERO_INV
- [ ] THRESH_OTSU
- [ ] THRESH_TRIANGLE

some of them could be implemented by [`torch.nn.functional.threshold`](https://pytorch.org/docs/stable/generated/torch.nn.functional.threshold.html), others from scratch.
______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

**Comment by erentknn** [comment]
I am going to start working on this if no one else didn't.

**Comment by edgarriba** [comment]
Sounds good!

**Comment by SeKim12** [comment]
Hi! I wanted to take a stab at this if nobody is currently working on this. Quick question -- would it be preferable for the thresholding to be differentiable (e.g. using sigmoid)? Thanks!

**Comment by edgarriba** [comment]
hi @SeKim12 if we can make it differentiably and keep it relatively fast would be great

**Comment by Big-Marvel** [comment]
Hi @edgarriba 
If this issue is available, I'd like to take this up.
Plan is to implement `cv::threshold` equivalent in `kornia.enhance` as a batched device-agnostic function.
I'd go with THRESH_BINARY, THRESH_BINARY_INV, THRESH_TRUNC, THRESH_TOZERO, THRESH_TOZERO_INV first.

**Comment by 0x-avi** [comment]
Hey @edgarriba i went through the functions of both libraries, and doesn't it make sense to use something like torch.where() ? it seems nn.Threshold is quite inflexible and i think adapting it to any of the cv2.threshold functions might be difficult . also is it okay if i implement all of them using just torch functions?

**Comment by edgarriba** [comment]
probably you are right

**Comment by shijianjian** [comment]
closed by #3515
